### PR TITLE
Add package icon and dependencies

### PR DESCRIPTION
**Pull request by dominichofer**

### CODE REVIEW

**Comment by Try2Code** [comment]
@dominichofer Can you explain a bit why adding a package (icon), which cannot be installed by the average spack user? Having a public package for a non-(public)-accessible software seems questionable at a first glance

**Comment by dominichofer** [comment]
Hi @skosukhin,
I'm trying to continue your excellent work by putting these packages into spack, so the world may use them.
I'm going to maintain icon as good as I can, but there are still parts that elude me and compilers I have no access to. AFAIK you are the expert in these regards, so would you help me maintain it?
For yaxt, libcdi-pio, rttov and scales-ppm, I don't see myself fit to maintain them. Would you be willing to adopt them yourself or do you know someone who might be a fit?

**Comment by dominichofer** [comment]
@Try2Code Certainly. I'm trying to help pushing the open source spirit. ICON is not public yet, but it's already used by the [German weather services](https://www.dwd.de/DE/Home/home_node.html), [MeteoSwiss](https://www.meteoschweiz.admin.ch/), the [Max Planck institute](https://mpimet.mpg.de/startseite), ETH Zürich et cetera on a number of supercomputers around the world. The community is working on making ICON public. Maybe by the end of the year or beginning of next.
Does that fit into Spack?

**Comment by Try2Code** [comment]
> @Try2Code Certainly. I'm trying to help pushing the open source spirit. ICON is not public yet, but it's already used by the [German weather services](https://www.dwd.de/DE/Home/home_node.html), [MeteoSwiss](https://www.meteoschweiz.admin.ch/), the [Max Planck institute](https://mpimet.mpg.de/startseite), ETH Zürich et cetera on a number of supercomputers around the world. The community is working on making ICON public. Maybe by the end of the year or beginning of next. Does that fit into Spack?

yh why not. Adding a package, that cannot be anonymously tested should not done for technical reasons IMO. Users should be able to fix it as much as possible. whenever the ICON code is available to the public, a package make sense of course. I have doubts that building a model with spack is a good choice for model developers, but this is a more personal preference I guess.

**Comment by dominichofer** [comment]
@Try2Code Valid point. I thought about this too. My reasoning is that spack supports packages which require a paid license https://spack.readthedocs.io/en/latest/build_systems/intelpackage.html#licensed-packages and packages that require manual downloading https://spack.readthedocs.io/en/latest/spack.html?highlight=manual%20download#spack.package_base.PackageBase.manual_download , thus ICON fits IMO. I don' t know if Spack will be the preferred way for model developers to install ICON. No one is forcing them to, but this package and access to the repo gives them this additional option. They could also just install the dependencies in a spack env with a view and then use ICON's buildsystem for building; Ulrich Schättler is currently pondering this.

**Comment by bernhardkaindl** [comment]
@Try2Code: @dominichofer is right. In addition, `spack` supports even HW platforms (Cray) which are not easy to come by and test.

Like he wrote, `spack` has recipes for packages you need registration (notably Intel software) for and manual download. I think some even may need even some keys to work.

Therefore, I see no reason why `spack` should not support installing ICON or/and its dependencies.

@Try2Code, are you fine with that?

If so, @dominichofer, would you add yourself and/or @Try2Code as maintainers of the dependencies?

@dominichofer please ping me or other like `alalzo` when ready for review/merge!

**Review by bernhardkaindl — COMMENTED** [review]
@Try2Code: @dominichofer is right. In addition, `spack` supports even HW platforms (Cray) which are not easy to come by and test.

Like he wrote, `spack` has recipes for packages you need registration (notably Intel software) for and manual download. I think some even may need even some keys to work.

Therefore, I see no reason why `spack` should not support installing ICON or/and its dependencies.

@Try2Code, are you fine with that?

If so, @dominichofer, would you add yourself and/or @Try2Code as maintainers of the dependencies?

@dominichofer please ping me or others like `alalzo` when ready for review/merge!

**Comment by Try2Code** [comment]
@bernhardkaindl @dominichofer 

IMO this MR is too big. Adding multiple packages is not needed, because the tar-ball of ICON does ship all the dependencies and the building knows all about them. 

Here is my wish list
1. Add only a single package ICON, which in a first version use the 2.6.6 tarball (locally downloaded, including its checksum) to build everything needed
2. Use  url pointing to where/how to obtain the code as long as the code is not freely accessible 
3. drop any in-necessary patches
4. Add new packages as replacements for the vendor-shipped ones step-by-step at a later point

I consider the support of the development-version hosted at gitlab.dkrz.de a bad idea at the moment, because the building has significantly changed from 2.6.6 and it will go like this in the future. ICON was and still is under constant development. For example: CLAW has been completely removed from the current development, but the package does handle all versions in the same way. Hence the package in its current state offers opportunities for severe building miss-configuration in my opinion. This could be avoided by limiting it to the 2.6.6 tar ball at first.

The ICON-building (dev version) already has options to account for externally installed versions of the shipped dependencies. esp. CLAW was difficult to handle. 2.6.7 is planned to be released this year, it might be the last non-open-access release. maybe it will be postponed to be the first Open-Access release for January 2024 - I am not 100% sure. Ether way the spackage will need further work to account for the differences between 2.6.6 and 2.6.7. 

This is another reason, why I'd like to start with a simple version of the icon package and work on it step by step. I don't see any problem from the user-side by limiting the 2.6.6 version to its shipped dependencies, but that's only my perspective.

As soon ICON is publicly available the package might be opened for the stable development branch directly from the repo. This would be a great way of distributing bug fixes to possible users. I really like the idea of having a package dedicated to the model itself (hope u didn't get the opposite impression from my previous posts). I happily volunteer to co-maintain such a thing.

**Comment by bernhardkaindl** [comment]
@Try2Code I agree on the basis that this MR is too big.

@dominichofer The usual guideline is that spack merges packages one by one:

The way to make progress would be to add the dependencies of icon (those that will be needed by the upcoming version) using one package with one PR.

If you both can support each other on this regular path forward, I think this would be good.

I'd let PR stand open, marked as draft (no further review) for reference to create the initial pull requests for adding the individual packages using separate pull requests.

Of course, you can create links between the new PRs and this PR by mentioning/(referring to) the other side in a comment if wanted.

@dominichofer If you agree, please set to draft (or close this PR).

**Comment by dominichofer** [comment]
> Here is my wish list
> 
> 1. Add only a single package ICON, which in a first version use the 2.6.6 tarball (locally downloaded, including its checksum) to build everything needed
> 2. Use  url pointing to where/how to obtain the code as long as the code is not freely accessible
> 3. drop any in-necessary patches
> 4. Add new packages as replacements for the vendor-shipped ones step-by-step at a later point

With this PR it's possible to dev-build local sources. Reducing the capabilities of this package to the tarball fells like a big step backwards for me. 
2. That's not what url is used for in spack.
3. There are no in-necessary patches.

> Ether way the spackage will need further work to account for the differences between 2.6.6 and 2.6.7.
This is another reason, why I'd like to start with a simple version of the icon package and work on it step by step. I don't see any problem from the user-side by limiting the 2.6.6 version to its shipped dependencies, but that's only my perspective.

Yes and whenever ICON changes the build system's API with a new version, the spackage needs to adapt. That's normal and expected.
Please feel free to create your own simple version of icon and work step by step, but I think this spackage is quite mature already.

File: var/spack/repos/builtin/packages/icon/package.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+class Icon(AutotoolsPackage, CudaPackage):
+    """Icosahedral Nonhydrostatic Weather and Climate Model."""
+
+    homepage = "https://code.mpimet.mpg.de/projects/iconpublic"
+    url = "https://gitlab.dkrz.de/icon/icon/-/archive/icon-2.6.6/icon-icon-2.6.6.tar.gz"
+    git = "git@gitlab.dkrz.de:icon/icon.git"
+
+    maintainers("dominichofer")
+
+    version("develop", submodules=True)
```
Comment (dominichofer) [inline var/spack/repos/builtin/packages/icon/package.py]:
```suggestion
```

**Comment by skosukhin** [comment]
I appreciate the interest in my work but I started this journey in October 2016 and would like to make the final step myself. Hopefully in not so distant future. Thank you.

**Comment by dominichofer** [comment]
@skosukhin I understand and respect this. Thus closing this PR.
If I can help you in any way, let me know.

**Pull request closed without merging**